### PR TITLE
Add T1168 and T1053 attack techniques (job scheduling)

### LINKED ISSUE #683: Add "Local Job Scheduling" attack technique (T1168)
[**T1168**](https://attack.mitre.org/techniques/T1168/)

Job scheduling can be used by adversaries to schedule running malicious code at some specified date and time.

**Adding it as a PBA:**
\- *LINUX*: There are multiple methods supported by Linux for creating pre-scheduled and periodic background jobs.
* _at_: not all systems have `at` so skip this (or check if it exists and then proceed?)  
* _cron_: attempt to add some file with some command (commented) to `/etc/cron.hourly` directory then delete it  

\- *WINDOWS*: not applicable ([T1053](https://attack.mitre.org/techniques/T1053/))

[**Mapping the technique to the ATT&CK matrix**](https://github.com/shreyamalviya/monkey/blob/guide-to-add-attack-technique/monkey/Adding-a-new-attack-technique.md)

### LINKED ISSUE #685: Add "Scheduled Task" attack technique (T1053)
[**T1053**](https://attack.mitre.org/techniques/T1053/)

Scheduled tasks can be used by adversaries to schedule running malicious code at some specified date and time.

**Adding it as a PBA:**
\- *LINUX*: not applicable ([T1168](https://attack.mitre.org/techniques/T1168/))
\- *WINDOWS*: using `schtasks` (see [this](https://github.com/redcanaryco/atomic-red-team/blob/master/atomics/T1053/T1053.md#atomic-test-2---scheduled-task-local))

[**Mapping the technique to the ATT&CK matrix**](https://github.com/shreyamalviya/monkey/blob/guide-to-add-attack-technique/monkey/Adding-a-new-attack-technique.md)

### PR DESCRIPTION
**Pull request by shreyamalviya**
Fixes #683 and fixes #685 

[T1168](https://attack.mitre.org/techniques/T1168/) and [T1053](https://attack.mitre.org/techniques/T1053/) both aim at scheduling jobs.

### CODE REVIEW

File: monkey/infection_monkey/post_breach/job_scheduling/windows/job_scheduling.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+def get_windows_commands_to_schedule_jobs():
+    return [
+        'schtasks',
+        '/Create',
+        '/SC',
+        'monthly',
+        '/TN',
+        SCHEDULED_TASK_NAME,
+        '/TR',
+        SCHEDULED_TASK_COMMAND
```
Comment (VakarisZ) [inline monkey/infection_monkey/post_breach/job_scheduling/windows/job_scheduling.py:14]: Not very readable. From human perspective this is a command, not an array. I think subprocess can take a string as a command. If not, then maybe lets return `" ".split(schtasks /Create /SC`

File: monkey/infection_monkey/post_breach/job_scheduling/windows/job_scheduling.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+def get_windows_commands_to_remove_scheduled_jobs():
+    return [
+        'schtasks',
+        '/Delete',
+        '/TN',
+        SCHEDULED_TASK_NAME,
+        '/F',
+        '>',
+        'nul',
+        '2>&1'
```
Comment (VakarisZ) [inline monkey/infection_monkey/post_breach/job_scheduling/windows/job_scheduling.py:27]: same as previous comment, could be refactored into a form, that we are used to reading.

File: monkey/monkey_island/cc/services/attack/technique_reports/T1053.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
+from monkey_island.cc.services.attack.technique_reports import AttackTechnique
+from monkey_island.cc.database import mongo
+from common.utils.attack_utils import ScanStatus
+from common.data.post_breach_consts import POST_BREACH_JOB_SCHEDULING
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/services/attack/technique_reports/T1053.py:4]: Ideally, all imports should follow the same structure:
```
future imports

python imports

third party imports

project imports
```
If not sure about why, look up PEP8.
You can automatically achieve this on pycharm with ctrl + alt + o.

File: monkey/infection_monkey/post_breach/job_scheduling/linux/job_scheduling.py [hunk 4]
```diff
@@ -0,0 +1,11 @@
+TEMP_CRON = "$HOME/monkey-schedule-jobs"
+
+
+def get_linux_commands_to_schedule_jobs():
+    return [
+        'touch {} &&'.format(TEMP_CRON),
+        'crontab -l > {} &&'.format(TEMP_CRON),
+        'echo \"# Successfully scheduled a job using crontab\" |',
+        'tee -a {} &&'.format(TEMP_CRON),
+        'crontab {}'.format(TEMP_CRON)
```
Comment (VakarisZ) [inline monkey/infection_monkey/post_breach/job_scheduling/linux/job_scheduling.py:10]: 1. Use python3 formatting `f"touch {TEMP_CRON} &&"`
2. you came up with this yourself? what tests did you do? Which command deletes `$HOME/monkey-schedule-jobs` file?

File: monkey/infection_monkey/post_breach/job_scheduling/linux/job_scheduling.py [hunk 5]
```diff
@@ -0,0 +1,11 @@
+TEMP_CRON = "$HOME/monkey-schedule-jobs"
```
Comment (VakarisZ) [inline monkey/infection_monkey/post_breach/job_scheduling/linux/job_scheduling.py:1]: no need to create a dir for a single file. It's better to do `job_scheduling_linux.py`. Not very important though

File: monkey/infection_monkey/post_breach/job_scheduling/job_scheduling.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+def get_commands_to_schedule_jobs():
+    linux_cmds = get_linux_commands_to_schedule_jobs()
+    windows_cmds = get_windows_commands_to_schedule_jobs()
+    return linux_cmds, windows_cmds
+
+
+def remove_scheduled_jobs():
+    subprocess.run(get_windows_commands_to_remove_scheduled_jobs() if is_windows_os()  # noqa: DUO116
+                   else '',
```
Comment (VakarisZ) [inline monkey/infection_monkey/post_breach/job_scheduling/job_scheduling.py:18]: This code give the impression that `subprocess.run('', shell=True)` removes scheduled job. If it doesn't then refactor into 
```
if is_windows_os():
    subprocess.run(get_windows_commands_to_remove_scheduled_jobs(), shell=True) 
```
There's no shame in splitting a mile long line that does everything into separate calls.

File: monkey/monkey_island/cc/services/attack/technique_reports/T1053.py (hunk 3 above)
Comment (ShayNehmad) [reply monkey/monkey_island/cc/services/attack/technique_reports/T1053.py:4]: Maybe we should add https://pypi.org/project/flake8-import-order/

File: monkey/infection_monkey/post_breach/job_scheduling/linux/job_scheduling.py (hunk 4 above)
Comment (shreyamalviya) [reply monkey/infection_monkey/post_breach/job_scheduling/linux/job_scheduling.py]: Yes, came up with it myself; updated - `$HOME/monkey-schedule-jobs` is deleted now.

File: monkey/infection_monkey/post_breach/job_scheduling/linux_job_scheduling.py [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+
+def get_linux_commands_to_schedule_jobs():
+    return [
+        f'touch {TEMP_CRON} &&',
+        f'crontab -l > {TEMP_CRON} &&',
+        'echo \"# Successfully scheduled a job using crontab\" |',
+        f'tee -a {TEMP_CRON} &&',
+        f'crontab {TEMP_CRON} &&',
+        f'rm {TEMP_CRON}'
+    ]
```
Comment (ShayNehmad) [inline monkey/infection_monkey/post_breach/job_scheduling/linux_job_scheduling.py:12]: We should `rm TEMP_CRON` even if the previous commands fail.

Comment (shreyamalviya) [reply monkey/infection_monkey/post_breach/job_scheduling/linux_job_scheduling.py]: Done; should I merge now?

**Pull request merged**